emma-language/src/main/scala/org/emmalanguage/api/DataBag.scala

```diff
@@ -311,6 +316,34 @@ trait DataBag[A] extends Serializable {
           }
       })._1
   }
+
+  // -----------------------------------------------------
+  // equals, hashCode and toString
+  // -----------------------------------------------------
+
+  override def equals(o: Any) = o match {
+    case that: DataBag[A] => {
+      val thisVals = fetch()
+      val thatVals = that.fetch()
+      (thisVals.size == thatVals.size) &&
```

Comment:
> The problem is that the user can actually use exists to implement minus

A definition such as

```scala
def minus(that: DataBag[A]): DataBag[A] = for {
  x <- xs; if ys.exists(x)
} yield x
```

opens the door for inconsistencies. For example with

```scala
def equals(that: DataBag[A]): Boolean =
  this.size == that.size && (this minus that).isEmpty && (that minus this).isEmpty
```

you will get wrong results.

```scala
DataBag(Seq(1,2,2,3)) equals DataBag(Seq(1,1,2,3)) 
```

To make `minus` compatible with the `equals` definition, it should be modified to account for the multiplicities. 

The more general problem is that defining a Boolean algebra on bags is not consistent with the usual semantics of `union` taken by the targeted backends, which is more efficient to implement and sufficient for most real-world scenarios.